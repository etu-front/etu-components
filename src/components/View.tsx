import React, { FC, CSSProperties } from 'react'
import styled from 'styled-components'
import { BaseProps } from './types'


const Container = styled.div`
  align-items: stretch;
  justify-content: flex-start;
  box-sizing: border-box;
  display: flex;
  position: relative;
  flex-direction: column;
`

type Align = 'center' | 'flex-start' | 'flex-end' | 'stretch' | 'space-between' | 'space-around'

interface IProps extends BaseProps {
  flex?: number | string
  align?: Align
  justify?: Align
  row?: boolean
  column?: boolean
  wrap?: boolean
  radius?: number | string
  height?: number | string
  background?: string
  color?: string
  width?: number | string
}

type ViewComponent = FC<IProps>

const View: ViewComponent & { Center: ViewComponent } = props => {
  const {
    className = '',
    children,
    align,
    justify,
    column = false,
    wrap = false,
    row = false,
    radius,
    height,
    background,
    color,
    flex,
    width,
    style,
    ...rest
  } = props

  let customStyle: CSSProperties = {}

  if (row) {
    customStyle.flexDirection = 'row'
    customStyle.alignItems = 'flex-start'
  }

  if (column) {
    customStyle.flexDirection = 'column'
  }

  if (align) {
    customStyle.alignItems = align
  }

  if (justify) {
    customStyle.justifyContent = justify
  }

  if (wrap) {
    customStyle.flexWrap = 'wrap'
  }

  if (flex) {
    customStyle.flex = flex
  }

  if (typeof width !== 'undefined') {
    customStyle.width = width
  }

  if (typeof height !== 'undefined') {
    customStyle.height = height
  }

  if (background) {
    customStyle.background = background
  }

  if (typeof color !== 'undefined') {
    customStyle.color = color
  }

  if (typeof radius !== 'undefined') {
    customStyle.borderRadius = radius
  }

  if (style) {
    customStyle = { ...customStyle, ...style }
  }

  return (
    <Container className={className} style={Object.keys(customStyle).length ? customStyle : undefined} {...rest}>
      {children}
    </Container>
  )
}
View.displayName = 'View'

View.Center = ({ children, ...rest }) => <View {...rest} align="center" justify="center">{children}</View>
View.Center.displayName = 'View.Center'

export default View
