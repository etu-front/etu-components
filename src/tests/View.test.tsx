import React from 'react'
import renderer from 'react-test-renderer'
import 'jest-styled-components'

import View from '../components/View'

describe('View test', () => {
  it('View Component', () => {
    expect(View.displayName).toEqual('View')
    expect(View.Center.displayName).toEqual('View.Center')
  })

  it('View 基本', () => {
    const component = renderer.create(<View column>hello</View>)
    const tree = component.toJSON()
    expect(tree).not.toBeNull()
    if (!tree) return
    expect(tree).toHaveStyleRule('align-items', 'stretch')
    expect(tree).toHaveStyleRule('flex-direction', 'column')
    expect(tree.props.style).toEqual({ "flexDirection": "column" })
    expect(tree.props.className).toContain(' ')
    expect(tree.children).toEqual(['hello'])
    expect(tree).toMatchSnapshot()
  })

  it('View 复杂样式', () => {
    const jsx = (
      <View
        row
        align={'flex-end'}
        flex={1}
        width={200}
        height={100}
        wrap
        justify={'stretch'}
        background={'red'}
        color={'white'}
        radius={5}
        style={{ lineHeight: '24px' }}
      >
        good
      </View>
    )
    const component = renderer.create(jsx)
    const tree = component.toJSON()
    expect(tree).not.toBeNull()
    if (!tree) return
    expect(tree.children).toEqual(['good'])
    const { style } = tree.props
    expect(style).toHaveProperty('flexDirection', 'row')
    expect(style).toHaveProperty('alignItems', 'flex-end')
    expect(style).toHaveProperty('flexWrap', 'wrap')
    expect(style).toHaveProperty('justifyContent', 'stretch')
    expect(style).toHaveProperty('flex', 1)
    expect(style).toHaveProperty('width', 200)
    expect(style).toHaveProperty('height', 100)
    expect(style).toHaveProperty('background', 'red')
    expect(style).toHaveProperty('color', 'white')
    expect(style).toHaveProperty('borderRadius', 5)
    expect(style).toHaveProperty('lineHeight', '24px')
    expect(tree).toMatchSnapshot()
  })

  it('View 样式', () => {
    const tree = renderer.create(
      <View wrap row style={{ margin: '0 -10px' }}>
        <View height={100} flex={1} background="red" />
        <View height={100} flex={1} background="green" />
        <View height={100} flex={1} background="blue" />
      </View>
    ).toJSON()
    expect(tree).toMatchSnapshot()
    const { style } = tree!.props
    expect(style).toHaveProperty('flexWrap', 'wrap')
    expect(style).toHaveProperty('flexDirection', 'row')
    expect(style).toHaveProperty('margin', '0 -10px')
  })

  it('View.Center 居中', () => {
    const tree = renderer.create(<View.Center height={200}>Hello</View.Center>).toJSON()
    expect(tree).not.toBeNull()
    if (!tree) return

    expect(tree).toMatchSnapshot()
    const { style } = tree.props
    expect(style).toHaveProperty('alignItems', 'center')
    expect(style).toHaveProperty('height', 200)
    expect(style).toHaveProperty('justifyContent', 'center')
  })
})
