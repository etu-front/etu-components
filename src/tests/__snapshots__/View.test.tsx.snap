// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`View test View 基本 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
  style={
    Object {
      "flexDirection": "column",
    }
  }
>
  hello
</div>
`;

exports[`View test View 复杂样式 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
  style={
    Object {
      "alignItems": "flex-end",
      "background": "red",
      "borderRadius": 5,
      "color": "white",
      "flex": 1,
      "flexDirection": "row",
      "flexWrap": "wrap",
      "height": 100,
      "justifyContent": "stretch",
      "lineHeight": "24px",
      "width": 200,
    }
  }
>
  good
</div>
`;

exports[`View test View 样式 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
  style={
    Object {
      "alignItems": "flex-start",
      "flexDirection": "row",
      "flexWrap": "wrap",
      "margin": "0 -10px",
    }
  }
>
  <div
    className="c0"
    style={
      Object {
        "background": "red",
        "flex": 1,
        "height": 100,
      }
    }
  />
  <div
    className="c0"
    style={
      Object {
        "background": "green",
        "flex": 1,
        "height": 100,
      }
    }
  />
  <div
    className="c0"
    style={
      Object {
        "background": "blue",
        "flex": 1,
        "height": 100,
      }
    }
  />
</div>
`;

exports[`View test View.Center 居中 1`] = `
.c0 {
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
  -webkit-box-pack: start;
  -webkit-justify-content: flex-start;
  -ms-flex-pack: start;
  justify-content: flex-start;
  box-sizing: border-box;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
  className="c0"
  style={
    Object {
      "alignItems": "center",
      "height": 200,
      "justifyContent": "center",
    }
  }
>
  Hello
</div>
`;
